[Merged by Bors] - refactor(algebra/group/defs): remove left-right_cancel_comm_monoids

adomani opened the pull request:
There were 6 distinct classes
* `(add_)left_cancel_comm_monoid`,
* `(add_)right_cancel_comm_monoid`,
* `(add_)cancel_comm_monoid`.

I removed them all, except for the last 2:
* `(add_)cancel_comm_monoid`.

The new typeclass `cancel_comm_monoid` requires only `a * b = a * c → b = c`, and deduces the other version from commutativity.

---

Please, can you *really* make sure that I have not messed up?  Thanks!  In particular, I have moved the `attribute [to_additive]` *after* the various typeclasses in question (UPD: Yury moved them back).  For more information, see the Zulip discussion.

Many thanks to Anne, Eric, Kevin and Mario for their explanations!  I have made an effort to understand them, but large parts of them are still mysterious to me!

Zulip discussion:
https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/mul_right_comm_monoid

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomani committed:
unfeat(algebra/group/defs): remove left-right_cancel_comm_monoids

Zulip discussion:
https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/mul_right_comm_monoid

adomani committed:
attributes

adomani committed:
add_comm_monoid swap

adomani committed:
still comm_monoid

adomani committed:
remove `add_` outside of `[to_additive]`

adomani committed:
explicitly mention `add_cancel_stuff`

Vierkantor commented on src/algebra/group/defs.lean:
```diff
@@ -305,18 +291,20 @@ class add_cancel_monoid (M : Type u)
-  extends add_left_cancel_comm_monoid M, add_right_cancel_comm_monoid M
+@[protect_proj, ancestor add_right_cancel_monoid add_left_cancel_monoid add_comm_monoid]
+class add_cancel_comm_monoid (M : Type u) extends
+  add_right_cancel_monoid M, add_left_cancel_monoid M, add_comm_monoid M
```
We should probably make these inherit from `add_cancel_monoid`, since `add_cancel_monoid` is the union of `add_left_cancel_monoid` and `add_right_cancel_monoid`.
```suggestion
@[protect_proj, ancestor add_cancel_monoid add_comm_monoid]
class add_cancel_comm_monoid (M : Type u) extends
  add_cancel_monoid M, add_comm_monoid M
```

Vierkantor commented on src/algebra/group/defs.lean:
```diff
@@ -305,18 +291,20 @@ class add_cancel_monoid (M : Type u)
-class cancel_comm_monoid (M : Type u) extends left_cancel_comm_monoid M, right_cancel_comm_monoid M
+@[protect_proj, ancestor right_cancel_monoid left_cancel_monoid comm_monoid]
+class cancel_comm_monoid (M : Type u) extends
+  right_cancel_monoid M, left_cancel_monoid M, comm_monoid M
```
```suggestion
@[protect_proj, ancestor cancel_monoid comm_monoid]
class cancel_comm_monoid (M : Type u) extends
  cancel_monoid M, comm_monoid M
```

adomani committed:
Apply suggestions from code review

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

urkud committed:
Move `@[to_additive]` back, fix def

urkud committed:
Move 1 more `to_additive` back

adomani committed:
algebra/ordered_monoid
remove `mul_right_cancel` from ordered_cancel_comm_monoid

adomani committed:
algebra/ordered_group
remove `mul_right_cancel` from linear_ordered_comm_group.to_linear_ordered_cancel_comm_monoid

adomani committed:
algebra/ordered_group
remove `mul_right_cancel` from ordered_comm_group.to_ordered_cancel_comm_monoid

adomani committed:
algebra/ordered_ring
remove several add_right_cancel

adomani committed:
data/nat/basic

urkud committed:
More fixes

urkud committed:
More fixes

urkud committed:
Fix lint
